# DATA2001 2020 S1 Tutorial Week 02
## Data Exploration with Python

## Exercise 1. Reading and Accessing a CSV File

We start by importing python libraries to read in a .csv file. We then visualise some rows of the data.

### Using Pandas

Pandas is a python data processing and analysis library that greatly eases data accessibility and usability to data sources. You can find more information in the [Week 02 lectures, from Page 15, slide 30](https://canvas.sydney.edu.au/courses/21543/files/9665910?module_item_id=759960).

Primarily we will be using pandas to load data from sources such as csv into it's DataFrame data-type.
A DataFrame is similar to a table in structure - it has rows indexed like an array. Each row has a list of fields. Each of those fields has a corresponding Heading.

We will be using the csv output of our Programming Experience Survey from Week 01.

Let's first import the relevant libraries that we require.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
Next we will import our CSV file named: 'programming_experience_survery_2020.csv'.

To do this in pandas we will use the 'read_csv(*filename*)' function and assigning the result to a new **DataFrame** variable.

Once we have imported the data, we can see the 'shape' of our data (how many rows and fields) to check whether what we have loaded is correct.

We can also use the 'head' operation of our DataFrame variable to see the first few rows of data.

In [31]:
# Import data
rawData = pd.read_csv('programming_experience_survey_2020.csv')
# print(rawData.dtypes)
# print(type(rawData["submitted"]))
print(rawData.shape)
rawData.head()

(263, 10)


respondent_id                submitted  \
0              1  2020-03-05 08:49:10 UTC   
1              2  2020-03-05 05:55:35 UTC   
2              3  2020-03-04 13:59:27 UTC   
3              4  2020-03-03 23:46:26 UTC   
4              5  2020-03-03 06:47:16 UTC   

                                      Tutorial Class  \
0  Friday 10 a.m. to 12 p.m., weeks 1-6 and 8-13 ...   
1                                                F12   
2                    Friday, 10-12 A.M., SIT lab 114   
3                                               F10D   
4                               Friday 2-4pm SIT 116   

                               Degree  Programming Experience  \
0            biology and data science                     1.0   
1              Biomedical engineering                     NaN   
2        Advanced Computing, 2rd year                     3.0   
3  Bachelor of Economics, second year                     6.0   
4      Bachelor of Project Management                     1.0   

                                Experience in Python Used Jupyter Notebooks  \
0                Written some simple Python programs                     No   
1                                               None                     No   
2      Have written complex Python programs already.                    Yes   
3  Competent Python programmer (familiar with eg....                     No   
4                                Basic understanding                     No   

  Other Programming Language Competency                     SQL Competency  \
0                                  Java    Heard of it, but never used it.   
1                              Matlab,R                Never heard of SQL.   
2                            C,C++,Java    Heard of it, but never used it.   
3                         Java,Matlab,R  Written some SQL queries already.   
4                                  Java  Written some SQL queries already.   

                    RDBMS Competency  
0     Microsoft Access,Oracle,Sybase  
1                   Microsoft Access  
2                             SQLite  
3                   Microsoft Access  
4  Microsoft SQL Server,MySQL,SQLite

## 1.1 TO DO for you Visualise the 9th, 15th, 31st, and 2nd last records

Display the 9th, 15th, 31th and the 2nd last record of the dictionary (hint: create an array of the appropriate indices annd use the DataFrame 'iloc' function to filter your DataFrame).

In [21]:
# TODO; 
#Double array to print multiple records
rawData.iloc[[9,15,31,-2]]

respondent_id                submitted  \
9               10  2020-02-29 07:56:17 UTC   
15              16  2020-02-28 05:19:46 UTC   
31              32  2020-02-28 03:24:38 UTC   
261            262  2020-02-19 23:51:57 UTC   

                             Tutorial Class  \
9                                    Fri AM   
15                                     F16A   
31   Friday 12-2pm \n\nSIT lab 114 (online)   
261                10am to 12pm SIT Lab 117   

                                                Degree  \
9                                                  NaN   
15   Bachelors of Software Engineering (Honours), 2...   
31   In the second year doing bachelor of Advanced ...   
261                  Bachelor of Economics, First year   

     Programming Experience  \
9                       3.0   
15                      1.0   
31                      1.0   
261                     0.0   

                                  Experience in Python Used Jupyter Notebooks  \
9                  Written some simple Python programs                    Yes   
15                 Written some simple Python programs                     No   
31   Competent Python programmer (familiar with eg....                     No   
261                Written some simple Python programs                     No   

    Other Programming Language Competency                      SQL Competency  \
9                                  C++,C#  Can interpret some SQL statements.   
15             Java,Javascript/ECMAScript  Can interpret some SQL statements.   
31                                 Java,R     Heard of it, but never used it.   
261       Other (answer below in comment)                 Never heard of SQL.   

                    RDBMS Competency  
9                                NaN  
15              Microsoft SQL Server  
31   Other (answer as comment below)  
261  Other (answer as comment below)

## 1.2 Rename dataframe columns

Usually when we receive data, the column/field labels are very long and descriptive. However, when we want to process data, we usually like to relabel columns/fields to make it easier to type.
:
First let's have a look at list of the column headings:

In [22]:
rawData.columns.values.tolist()

['respondent_id',
 'submitted',
 'Tutorial Class',
 'Degree',
 'Programming Experience',
 'Experience in Python',
 'Used Jupyter Notebooks',
 'Other Programming Language Competency',
 'SQL Competency',
 'RDBMS Competency']

Now let's create a working copy of our data so that we don't modify our original data. This a common practice to make sure we do not accidently corrupt our initial input data.

Once we have our working copy and our above list of headings, it's a simple matter of using the DataFrame 'rename(*rename_target*=*mapping dictionary of old to new field names*, inplace=True)' function to change the column/field headers within our working copy. The 'inplace=True' means modify the existing items in the DataFrame.

In [33]:
# Create a working copy of the raw data
wrkData = rawData.copy()

# Rename columns
wrkData.rename(columns={
    'respondent_id': 'ResId',
    'submitted': 'SubmitDateTime',
    'Tutorial Class': 'ClassId',
#     'Degree',
    'Programming Experience': 'PrgExp',
    'Experience in Python': 'PyExp',
    'Used Jupyter Notebooks': 'JupyterExp',
    'Other Programming Languages Competency': 'OtherExp',
    'SQL Competency': 'SQLExp',
    'Relational Databases Competency': 'RelDBExp'
}, inplace=True)


ResId                                      int64
SubmitDateTime                            object
ClassId                                   object
Degree                                    object
PrgExp                                   float64
PyExp                                     object
JupyterExp                                object
Other Programming Language Competency     object
SQLExp                                    object
RDBMS Competency                          object
dtype: object


Since we are loading from CSV, we do not have type association for our columns/fields (in fact, everything is considered to be a string). Type information is important because it can alter how the same operations affect the same data but with different types - consider a string vs a number vs a date:

   **"3" + "4" = "34"**
   
   **3 + 4 = 7**
   
   **3s + 4h = 4:00:03**
   
Let's try converting the 'SubmitDateTime' column/field to a date type and view our data again:

In [24]:
# Convert to datetime
# SubmitDateTime here is original "submitted" in rawData
wrkData['SubmitDateTime'] = pd.to_datetime(wrkData['SubmitDateTime'])

# Check df column data types
print(wrkData.dtypes)

# View
wrkData.head()

ResId                                                  int64
SubmitDateTime                           datetime64[ns, UTC]
ClassId                                               object
Degree                                                object
PrgExp                                               float64
PyExp                                                 object
JupyterExp                                            object
Other Programming Language Competency                 object
SQLExp                                                object
RDBMS Competency                                      object
dtype: object


ResId            SubmitDateTime  \
0      1 2020-03-05 08:49:10+00:00   
1      2 2020-03-05 05:55:35+00:00   
2      3 2020-03-04 13:59:27+00:00   
3      4 2020-03-03 23:46:26+00:00   
4      5 2020-03-03 06:47:16+00:00   

                                             ClassId  \
0  Friday 10 a.m. to 12 p.m., weeks 1-6 and 8-13 ...   
1                                                F12   
2                    Friday, 10-12 A.M., SIT lab 114   
3                                               F10D   
4                               Friday 2-4pm SIT 116   

                               Degree  PrgExp  \
0            biology and data science     1.0   
1              Biomedical engineering     NaN   
2        Advanced Computing, 2rd year     3.0   
3  Bachelor of Economics, second year     6.0   
4      Bachelor of Project Management     1.0   

                                               PyExp JupyterExp  \
0                Written some simple Python programs         No   
1                                               None         No   
2      Have written complex Python programs already.        Yes   
3  Competent Python programmer (familiar with eg....         No   
4                                Basic understanding         No   

  Other Programming Language Competency                             SQLExp  \
0                                  Java    Heard of it, but never used it.   
1                              Matlab,R                Never heard of SQL.   
2                            C,C++,Java    Heard of it, but never used it.   
3                         Java,Matlab,R  Written some SQL queries already.   
4                                  Java  Written some SQL queries already.   

                    RDBMS Competency  
0     Microsoft Access,Oracle,Sybase  
1                   Microsoft Access  
2                             SQLite  
3                   Microsoft Access  
4  Microsoft SQL Server,MySQL,SQLite

## Exercise 2. Data Cleaning and Conversion

It's often the case that form data is collected in terms of text data, but we wish to convert this to a numerical value. A common practice among questions that ascertain some sort of increasing level of respondance is to convert to a Likert scale - this is an escalating series of numbers that replaces each level of respondance.

### 2.1 Map Columns values to digits
Map the {PythonExp} column values as follows:
- 'None' = 1
- 'Basic understanding' = 2
- 'Written some simple Python programs' = 3
- 'Competent Python programmer (familiar with eg. functions and classes)' = 4
- 'Have written complex Python programs already.' = 5

Let's examine the particular responses we received for 'Experience in Python'.

This can be achieved using the DataFrame 'groupby' and 'size' functions.

In [45]:
# Get distinct list of {PythonExp} values
# wrkData.groupby('PyExp').size().reset_index(name='NumObs') # df
wrkData.groupby('PyExp').size() # Series
# wrkData.groupby('PyExp').agg('count')

PyExp
Basic understanding                                                       32
Competent Python programmer (familiar with eg. functions and classes)    121
Have written complex Python programs already.                             25
None                                                                       6
Written some simple Python programs                                       79
dtype: int64

### TODO for you - Map column values

Next we have to use our mapping established above to introduce a new field/column that contains the Likert scale values for our Python experience. You can label the new column/field 'PyExpLikert'.

Hint: Use the DataFrame command 'map(*mapping dictionary of old to new field values*)'

In [48]:
# TODO - Add a new field / column called 'PyExpLikert' that changes the Experience in Python responses to a Likert Scale
wrkData["PyExp"].map({"None" : 1, "Basic understanding" : 2, "Written some simple Python programs" : 3, "Competent Python programmer (familiar with eg. functions and classes)" : 4, "Have written complex Python programs already." : 5})

0      3
1      1
2      5
3      4
4      2
5      4
6      4
7      3
8      4
9      3
10     5
11     4
12     2
13     2
14     3
15     3
16     1
17     4
18     4
19     4
20     3
21     3
22     4
23     3
24     2
25     4
26     3
27     4
28     2
29     3
      ..
233    2
234    4
235    5
236    4
237    4
238    3
239    4
240    3
241    4
242    3
243    4
244    5
245    4
246    4
247    3
248    3
249    4
250    4
251    4
252    5
253    4
254    4
255    3
256    3
257    5
258    4
259    3
260    4
261    3
262    3
Name: PyExp, Length: 263, dtype: int64

### 2.2 Number of Other Languages Experience

Define the cardinality for each respondent by considering responses in regards to their experience described in 'Other programming languages competency'. 

For example, your response mapping should work as follows:

'C#, Java, Javascript/ECMAScript, Matlab' = 4

'Matlab,R' = 2 

'Haskell' = 1

... and so on. 

### TODO by you - Identify the number of other languages {OtherExpCnt} for each respondent. 

Add a new field/column to your working data ('OtherExpCnt') that will contain the converted cardinality values calculated above. 

Hint: consider the number of comma separated values in each cell. Try using the 'count(*character*)' property of strings in addition to the DataFrame mapping ability. 


In [ ]:
# TODO by you - Identify the number of other languages {OtherExpCnt} for each respondent. 



### 2.3 TODO by you - Number of Relational Database Competency {RelDBExp}

We also wish to identify the relation database experience as a numerical value as well. In a manner similar to the task above, identify the number of relational database compentencies and store it in a new field/column {RelDBExp} in our working data.

In [ ]:
# TODO by you - Number of Relational Database Competency {RelDBExp}



### 2.4 TODO by you - Encode SQL Competency {SQLExpLikert}

Similar to 2.1, encode a new field/column {SQLExpLikert} using the following encoding for the Likert scale of the 'SQLExp' field/column:
- 'Never heard of SQL' = 1
- 'Heard of it, but never used it' = 2
- 'Can interpret some SQL statements' = 3
- 'Written some SQL queries already' = 4
- 'Can already create tables and write complex SQL queries' = 5

First check distinct values of {SQLExp}.

In [ ]:
# wrkData['SQLExp'].unique().tolist()

In [ ]:
# TODO by you - Encode SQL Competency {SQLExpLikert}


### 2.5 (ADV only) TODO by you - Dealing with cleaning many varied strings

Define a Python method to clean the 'Tutorial class' column by analysing the texts in cells as follows (it will be written in all kinds of variants, the the following as a general mapping):

- SIT Lab 117 (Friday 8am-10am) = F08A
- SIT Lab 116 (Friday 8am-10am) = F08B
- SIT Lab 118 (Friday 8am-10am) = F08C
- SIT Lab 115 (Friday 8am-10am) (online) = F08D
- SIT Lab 114 advanced (Friday 8am-10am) = F08ADV
- SIT Lab 118 (Friday 10am-12pm) (online) = F10A
- SIT Lab 117 (Friday 10am-12pm) = F10B
- SIT Lab 116 (Friday 10am-12pm) = F10C
- SIT Lab 115 (Friday 10am-12pm) = F10D
- SIT Lab 114 advanced (Friday 10am-12pm) = F10ADV
- SIT Lab 115 (Friday 12-2pm) = F12A
- SIT Lab 114 (Friday 12-2pm) (online) = F12A
- SIT Lab 116 (Friday 12-2pm) = F12C
- SIT Lab 117 (Friday 12-2pm) = F12D
- SIT Lab 115 (Friday 2-4pm) = F14A
- SIT Lab 117 (Friday 2-4pm) = F14B
- SIT Lab 116 (Friday 2-4pm) = F14C
- SIT Lab 114 (Friday 12-2pm) (online) = F14D
- SIT Lab 115 (Friday 4-6pm) = F16A

All strings need to be cleaned and converted such that this column contains only the given (F08A/B/C/D, F10A/B/C/D, F12A/B/C/D, F14A/B/C/D, F16A) codes. All values which do not fit into this pattern (e.g. a single ”Yes” answer) should be converted to 'NA'.

You can begin with the cleaning function provided in the lectures. Note that this function changes data in place:

In [ ]:
def clean(data, column_key, convert_function, default_value, special_values):
#     special_values= {} # we provide the special values
    for row in data:
        old_value = row[column_key]
        new_value = default_value
        try:
            if old_value in special_values.keys():
                new_value = special_values[old_value]
            else:
                new_value = convert_function(old_value)
        except (ValueError, TypeError):
            print('Replacing {} with {} in column {}'.format(row[column_key], new_value, column_key))
        row[column_key] = new_value

## Exercise 3. Analysing Date and Time

Usually we need to do very specific operations on dates and times. These will usually involve seperating particular components or calculating durations.

### 3.1 TODO by you - Extract the Date {SubmitDate} and Time {'SubmitTime} components of the Submitted column/field

Create a new field/column {SubmitDate} for the Date component and a new field/column {SubmitTime} for the Time component from the submission time.

In [ ]:
# TODO by you - Extract the Date {SubmitDate}
# Get date component


In [ ]:
# TODO by you - Extract the Time {SubmitDate}
# Get time component


## Exercise 4. Data Visualisations

Data visualisation are an important part of analysing data and gaining insights. 

In this exercise, you must use a frequency plot and a box plot to visualise the data we have gathered and cleaned.

### 4.1 Frequency Plot / Histogram

Show a frequency plot of the 'Python Experience' as given by the respondents. In this plot, the x-axis should contain the Python Experience in level (1, 2, 3, 4, 5...) and y-axis should contain how many respondents there were for each level. Bar chart plotting reference: https://pythonspot.com/matplotlib-bar-chart/

In [ ]:
%matplotlib inline


### 4.2 Boxplots for Likert-Scale

Using boxplots this time, visualise boxplots for {OtherExpCnt} and {RelDBExpCnt}.

In [ ]:
%matplotlib inline


## Exercise 5. (Adv only) Integrating Unix into Jupyter Notebook

The general idea of the advanced exercise is to solve a few of the data cleaning and analysis steps with the help of some common Unix commands in some of thenotebook cells (cf. advanced seminar this week).

### 5.1 TODO by you - Python compentency

Determine  the  number  of  responses  for  each  Python  programming  competency level using Unix commands.

In [ ]:
# TODO by you - use Unix commands to determine the number of competency for each level of python programming

### 5.2 TODO by you - Other languages competency

Split the responses for additional language skills into individual language answers, and then determine the number of mentionings of each programming language (using at least in part Unix commands).

In [ ]:
# TODO by you - split the additional language skills into individual language answers, 
#  then determine the mentions of each programming language. Look at the wordcount example for inspiration


### 5.3 TODO by you - Visualise the results

Visual the results from the previous sub-question in a histogram plot. 

Which languages are most known?

Which languages are least known?

In [ ]:
# TODO by you - visualise the results of 5.2, you will have to save the results back into a python variable

### 5.4 TODO by you - Do the same for the databases skills. 

Which database systems are known/used most by students of this class?

In [ ]:
#TODO by you - repeat 5.2 and 5.3 for the different types of databases systems or skills that students know